include/nanorange/algorithm/stable_partition.hpp

```diff
@@ -95,50 +97,52 @@ struct stable_partition_fn {
         }
 
         const I first_false = (m1 == first) ? first :
-                impl_unbuffered(first, m1, len_half, pred, proj);
+                impl_unbuffered(first, m1, len_half, pred, proj).begin();
 
         m1 = middle;
         len_half = dist - half;
 
         while (nano::invoke(pred, nano::invoke(proj, *m1))) {
             if (++m1 == last) {
-                return nano::rotate(first_false, middle, ++last).begin();
+                auto rot = nano::rotate(first_false, middle, ++last);
+                return {std::move(rot.begin()), nano::next(last)};
             }
         }
 
-        const I last_false = impl_unbuffered(m1, last, len_half, pred, proj);
+        const I last_false = impl_unbuffered(m1, last, len_half, pred, proj).begin();
 
-        return nano::rotate(first_false, middle, last_false).begin();
+        auto rot = nano::rotate(first_false, middle, last_false);
+        return {std::move(rot.begin()), nano::next(last)};
     }
 
     template <typename I, typename Pred, typename Proj>
-    static I impl(I first, I last, Pred& pred, Proj& proj)
+    static subrange<I> impl(I first, I last, Pred& pred, Proj& proj)
     {
         // Find the first non-true value
         first = nano::find_if_not(std::move(first), last, std::ref(pred), std::ref(proj));
         if (first == last) {
-            return first;
+            return {std::move(first), std::move(last)};
         }
 
         // Find the last true value
-        last = nano::find_if(nano::make_reverse_iterator(last),
+        I it = nano::find_if(nano::make_reverse_iterator(last),
```

Comment:
Changed to `it` because we need the original `last` for the return value. That also means that we can't just `return impl_unbuffered(...)` below, since it is operating on `it`, rather than `last`. Same goes for `impl_buffered(...)`.

Reply:
Okay